clang/lib/CConv/ConstraintBuilder.cpp

```diff
@@ -527,27 +449,58 @@ class GlobalVisitor : public RecursiveASTVisitor<GlobalVisitor> {
   ASTContext *Context;
   ProgramInfo &Info;
   ConstraintResolver CB;
-  TypeVariableBindingsMapT &TypeVariableBindings;
+  TypeVarInfo &TVInfo;
 };
 
-// Store type param bindings persistently in ProgramInfo so they are available
-// during rewriting.
-void ConstraintBuilderConsumer::SetProgramInfoTypeVars
-   (TypeVariableBindingsMapT TypeVariableBindings, ASTContext &C) {
-  for (const auto &TVEntry : TypeVariableBindings) {
-    bool AllNull = true;
-    for (auto TVCallEntry : TVEntry.second)
-      AllNull &= TVCallEntry.second == nullptr;
-    if (!AllNull) {
-      for (auto TVCallEntry : TVEntry.second)
-        if (TVCallEntry.second != nullptr) {
-          std::string TyStr = TVCallEntry.second->getPointeeType().getAsString();
-          Info.setTypeParamBinding(TVEntry.first, TVCallEntry.first, TyStr, &C);
-        } else
-          Info.setTypeParamBinding(TVEntry.first, TVCallEntry.first, "void", &C);
+// Some information about variables in the program is required by the type
+// variable analysis and constraint building passes. This is gathered by this
+// visitor which is executed before both of the other visitors.
+class VariableAdderVisitor : public RecursiveASTVisitor<VariableAdderVisitor> {
+public:
+  explicit VariableAdderVisitor(ASTContext *Context, ProgramInfo &I)
+      : Context(Context), Info(I), CB(Info, Context) {}
+
+  bool VisitVarDecl(VarDecl *D) {
+    FullSourceLoc FL = Context->getFullLoc(D->getBeginLoc());
+    if (FL.isValid() &&
+        (!isa<ParmVarDecl>(D) || D->getParentFunctionOrMethod() != nullptr))
+      addVariable(D);
+    return true;
+  }
+
+  bool VisitFunctionDecl(FunctionDecl *D) {
+    FullSourceLoc FL = Context->getFullLoc(D->getBeginLoc());
+    if (FL.isValid())
+      Info.addVariable(D, Context);
+    return true;
+  }
+
+  bool VisitRecordDecl(RecordDecl *Declaration) {
+    if (RecordDecl *Definition = Declaration->getDefinition()) {
+      FullSourceLoc FL = Context->getFullLoc(Definition->getBeginLoc());
+      if (FL.isValid()) {
+        SourceManager &SM = Context->getSourceManager();
+        FileID FID = FL.getFileID();
+        const FileEntry *FE = SM.getFileEntryForID(FID);
+        if (FE && FE->isValid())
+          for (auto *const D : Definition->fields())
+            addVariable(D);
```

Comment:
Not `Info.addVariable` -- should the ones above call `addVariable` too, like this one is?

Reply:
The exact handling of these cases was different in the original code before I pulled it out into this new visitor. Not sure I fully understand all of what's going on.

Reply:
Oh that's interesting. I think this might be stuff that @sroy4899 worked on; maybe she can explain.

Reply:
From what I remember of this when I worked on it, it used to be Info.addVariable here too :/